### ISSUE TITLE
Backup and restore mogodb

### ISSUE THREAD

**Issue by Jungack** [overleaf/overleaf#660]
Hello,
I tried to setup Overleaf and all went well on my test machine ! But before installing on my main computer, I kept playing on the test machine.

Here we are : I just wanted to test my abilities concerning backup and restore the mongodb. So here comes the problem : I successfully made a backup using this command : 

`docker-compose exec -T mongo sh -c 'mongodump --archive' > db.dump`

Just after, I created a new project. And finally, I wanted to restore the old backup using : 

`docker-compose exec -T mongo sh -c 'mongorestore --archive' < db.dump`

This did not work since I always got the error : 
`E11000 duplicate key error collection`

I achieved to bypass It with the --drop argument : 
`docker-compose exec -T mongo sh -c 'mongorestore --archive --drop' < db.dump`

However, from now, the compiler does not work anymore. So what is wrong with what I did? I don't understand why this happened 😅...

**Comment by marcuslindner** [comment]
Hi,

We have a similar problem, which I believe is related. I tried to update our ShareLaTeX installation (v.1.2.1) to the new Overleaf (v2.0.1). We followed the upgrade procedure and essentially just pulled the new docker image and started a new container with the same settings as the old one.

The docker logs show me an error/warning and when logging in to the new Overleaf, the editor is not working. It never finishes loading and at some an error message appears.

This is the output of the docker logs:
![image](https://user-images.githubusercontent.com/8193090/68069671-a1d21480-fd63-11e9-9605-f6a4b8e32969.png)

**Comment by Jungack** [comment]
We don't have the same issue. In fact, the error log I can find about not compiling is in web.log. It shows this : 
`{"name":"web","hostname":"49c11d0ad6ce","pid":420,"level":30,"doc_id":"5dbcbcff66ad0e01a2b1c703","project_id":"5dbcbcff66ad0e01a2b1c702","version":5,"rev":2,"msg":"got doc from docstore api","time":"2019-11-02T10:54:25.046Z","v":0}
{"name":"web","hostname":"49c11d0ad6ce","pid":420,"level":30,"req":{"url":"/project/5dbcbcff66ad0e01a2b1c702/doc/5dbcbcff66ad0e01a2b1c703/metadata","method":"POST","remote-addr":"80.215.80.29","user-agent":"Mozilla/5.0 (Linux; Android 9; MI 8) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.90 Mobile Safari/537.36","content-length":"48"},"res":{"statusCode":200},"response-time":47,"msg":"http request","time":"2019-11-02T10:54:25.049Z","v":0}
{"name":"web","hostname":"49c11d0ad6ce","pid":420,"level":50,"err":{"message":"CLSI compile failed: error sending request to clsi: failed to make request to CLSI","name":"OError","stack":"OError: CLSI compile failed: error sending request to clsi: failed to make request to CLSI\n    at ClsiManager._sendBuiltRequest.result (/var/www/sharelatex/web/app/src/Features/Compile/ClsiManager.js:82:15)\n    at ClsiManager._postToClsi (/var/www/sharelatex/web/app/src/Features/Compile/ClsiManager.js:213:22)\n    at ClsiManager._makeRequest (/var/www/sharelatex/web/app/src/Features/Compile/ClsiManager.js:435:16)\n    at Request.request [as _callback] (/var/www/sharelatex/web/app/src/Features/Compile/ClsiManager.js:276:24)\n    at self.callback (/var/www/sharelatex/web/node_modules/request/request.js:186:22)\n    at Request.emit (events.js:198:13)\n    at Request.EventEmitter.emit (domain.js:448:20)\n    at Request.onRequestError (/var/www/sharelatex/web/node_modules/request/request.js:878:8)\n    at ClientRequest.emit (events.js:198:13)\n    at ClientRequest.EventEmitter.emit (domain.js:448:20)\n    at Socket.socketOnEnd (_http_client.js:426:9)\n    at Socket.emit (events.js:203:15)\n    at Socket.EventEmitter.emit (domain.js:448:20)\n    at endReadableNT (_stream_readable.js:1145:12)\n    at process._tickCallback (internal/process/next_tick.js:63:19)\n\ncaused by: OError: error sending request to clsi: failed to make request to CLSI\n    at ClsiManager._postToClsi (/var/www/sharelatex/web/app/src/Features/Compile/ClsiManager.js:214:17)\n    at ClsiManager._makeRequest (/var/www/sharelatex/web/app/src/Features/Compile/ClsiManager.js:435:16)\n    at Request.request [as _callback] (/var/www/sharelatex/web/app/src/Features/Compile/ClsiManager.js:276:24)\n    at self.callback (/var/www/sharelatex/web/node_modules/request/request.js:186:22)\n    at Request.emit (events.js:198:13)\n    at Request.EventEmitter.emit (domain.js:448:20)\n    at Request.onRequestError (/var/www/sharelatex/web/node_modules/request/request.js:878:8)\n    at ClientRequest.emit (events.js:198:13)\n    at ClientRequest.EventEmitter.emit (domain.js:448:20)\n    at Socket.socketOnEnd (_http_client.js:426:9)\n    at Socket.emit (events.js:203:15)\n    at Socket.EventEmitter.emit (domain.js:448:20)\n    at endReadableNT (_stream_readable.js:1145:12)\n    at process._tickCallback (internal/process/next_tick.js:63:19)\n\ncaused by: OError: failed to make request to CLSI\n    at ClsiManager._makeRequest (/var/www/sharelatex/web/app/src/Features/Compile/ClsiManager.js:436:11)\n    at Request.request [as _callback] (/var/www/sharelatex/web/app/src/Features/Compile/ClsiManager.js:276:24)\n    at self.callback (/var/www/sharelatex/web/node_modules/request/request.js:186:22)\n    at Request.emit (events.js:198:13)\n    at Request.EventEmitter.emit (domain.js:448:20)\n    at Request.onRequestError (/var/www/sharelatex/web/node_modules/request/request.js:878:8)\n    at ClientRequest.emit (events.js:198:13)\n    at ClientRequest.EventEmitter.emit (domain.js:448:20)\n    at Socket.socketOnEnd (_http_client.js:426:9)\n    at Socket.emit (events.js:203:15)\n    at Socket.EventEmitter.emit (domain.js:448:20)\n    at endReadableNT (_stream_readable.js:1145:12)\n    at process._tickCallback (internal/process/next_tick.js:63:19)\n","info":{"projectId":"5dbcbcff66ad0e01a2b1c702","userId":"5daf6dd48acb7200831e481f","compileOptions":{"compiler":"pdflatex","timeout":180,"draft":false,"check":"silent","syncType":"full","syncState":"41c4f59ad97ab2defaa28b3e197def99229ba6b7"},"rootResourcePath":"main.tex"}},"url":"/project/5dbcbcff66ad0e01a2b1c702/compile?auto_compile=true","method":"POST","user":{"id":"5daf6dd48acb7200831e481f","email":"crackk136@gmail.com"},"msg":"error passed to top level next middleware","time":"2019-11-02T10:54:40.180Z","v":0}
{"name":"web","hostname":"49c11d0ad6ce","pid":420,"level":30,"req":{"url":"/project/5dbcbcff66ad0e01a2b1c702/compile?auto_compile=true","method":"POST","remote-addr":"80.215.80.29","user-agent":"Mozilla/5.0 (Linux; Android 9; MI 8) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.90 Mobile Safari/537.36","content-length":"131"},"res":{"statusCode":500},"response-time":600446,"msg":"http request","time":"2019-11-02T10:54:40.276Z","v":0}
{"name":"web","hostname":"49c11d0ad6ce","pid":420,"level":30,"rss":"130.41","heapTotal":"76.89","heapUsed":"67.25","external":"34.29","msg":"process.memoryUsage()","time":"2019-11-02T10:55:22.073Z","v":0}
{"name":"web","hostname":"49c11d0ad6ce","pid":420,"level":30,"ip":"80.215.80.29","opts":{"url":"http://localhost:8080/json/80.215.80.29","timeout":1000,"json":true},"msg":"getting geo ip details","time":"2019-11-02T10:55:33.717Z","v":0}
{"name":"web","hostname":"49c11d0ad6ce","pid":420,"level":40,"err":{"message":"connect ECONNREFUSED 127.0.0.1:8080","name":"Error","stack":"Error: connect ECONNREFUSED 127.0.0.1:8080\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1106:14)\n","info":{},"code":"ECONNREFUSED"},"ip":"80.215.80.29","msg":"error getting ip details","time":"2019-11-02T10:55:33.718Z","v":0}
{"name":"web","hostname":"49c11d0ad6ce","pid":420,"level":30,"req":{"url":"/editingSession/5dbcbcff66ad0e01a2b1c702","method":"PUT","remote-addr":"80.215.80.29","user-agent":"Mozilla/5.0 (Linux; Android 9; MI 8) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.90 Mobile Safari/537.36","content-length":"0"},"res":{"statusCode":204},"response-time":12,"msg":"http request","time":"2019-11-02T10:55:33.721Z","v":0}
{"name":"web","hostname":"49c11d0ad6ce","pid":420,"level":50,"err":{"message":"connect ECONNREFUSED 127.0.0.1:3050","name":"Error","stack":"Error: connect ECONNREFUSED 127.0.0.1:3050\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1106:14)\n","info":{},"code":"ECONNREFUSED"},"url":"http://localhost:3050/editingSession","query":{"userId":"5daf6dd48acb7200831e481f","projectId":"5dbcbcff66ad0e01a2b1c702"},"msg":"Background request failed","time":"2019-11-02T10:55:44.172Z","v":0}
{"name":"web","hostname":"49c11d0ad6ce","pid":420,"level":50,"err":{"message":"connect ECONNREFUSED 127.0.0.1:3050","name":"Error","stack":"Error: connect ECONNREFUSED 127.0.0.1:3050\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1106:14)\n","info":{},"code":"ECONNREFUSED"},"url":"http://localhost:3050/editingSession","query":{"userId":"5daf6dd48acb7200831e481f","projectId":"5dbcbcff66ad0e01a2b1c702"},"msg":"Background request failed","time":"2019-11-02T10:56:20.273Z","v":0}
`
I have a CLSI failed and just after connection refused...

**Comment by marcuslindner** [comment]
No, not the same issue, but a related one. I think the root cause might be the same. You mention the E11000 error after restoring the database. We also get an E11000 error when the new installation tries to update the old database.

Somehow, both error are related to the database and that's why I think they might be related.

**Comment by mserranom** [comment]
Thanks for the report. These 2 look unrelated, since @Jungack manages to see the editor. 

I assume you're using Overleaf CE and not Server Pro, let me know otherwise.

- @Jungack the error is unrelated to the DB. You see the `E11000` because you're trying to load DB entries in the same database you jus backedup, using the `--drop` is correct. Can you share the content of `/var/log/sharelatex/clsi.log`?
- @marcuslindner seems you're having a problem when creating an index during the migration. For some reason you have 2 `null` entries for `tokens.readAndWrite` in `projects` collection, which is strange. This is what you can try (after making sure you've backed up your installation):
    - Login to mongo: `docker-compose exec mongo mongo` and switch db with `use sharelatex`.
    - Check that those `null` entries exist: `db.projects.find({'tokens.readAndWrite': {$exists: true, $eq: null}},{'tokens.readAndWrite':1})`
    - Remove the `null` entries: `db.projects.updateMany({'tokens.readAndWrite': {$exists: true, $eq: null}}, {$unset: 'tokens.readAndWrite'})`
    - Run the upgrade procedure and check the migration scripts go as expected.

**Comment by mserranom** [comment]
Another useful detail: could you confirm that you're both using dockerhub images and not custom builds?

**Comment by marcuslindner** [comment]
Hi mserranom,

yes, we're using the docker hub images (CE version).

Regarding the null entries, I tried to execute the commands to find (and potentially remove) them. Mongo complains that the 'eq' operator doesn't exist. I suppose my mongodb version is too old for that command? But it is the newest version that comes with Debian 8. Since we try to update our old installation, we have the MongoDB as local installation and not as docker container.

This is the error that I get:
![image](https://user-images.githubusercontent.com/8193090/68250781-fa363a00-0021-11ea-88dd-450c9d143546.png)

Is there perhaps an equivalent command that doesn't use the $eq operator?

**Comment by mserranom** [comment]
You can try:

```
db.projects.find({'tokens.readAndWrite': null},{'tokens.readAndWrite':1})
```

It will also print any project that doesn't have `tokens.readAndWrite` set, and then try to grep `null` values.

**Comment by ShaneKilkelly** [comment]
The `$eq` operator is new in MongoDB v3, which is the version we use for our services in production. In general we don't actively support any previous version of Mongo, so we recommend keeping up to date with recent releases. This should minimize divergence between your own installation and our production systems :)

**Comment by marcuslindner** [comment]
Yes, I agree. I can also not delete the entries with ".updateMany" and "$unset", since these two functions don't exist either in our old database.

I'm trying now to figure out how to update the database (backup and restore the data into the new database) and see what I can achieve then.

**Comment by ShaneKilkelly** [comment]
In the short term you could just use the default `update()` method and the `$unset` operation, it should be available in older versions of mongo.

Something like:

```
db.projects.update({_id: ObjectId("...")}, {$unset: 'tokens.readAndWrite'})
```

... and then run that once for each project id you found with this problem.

**Comment by marcuslindner** [comment]
I, again, agree to you very much. That's also my plan to move everything into docker containers and have a proper docker-compose configuration.

Currently, I face another problem. When I try to manually remove the null entries the way you described, I get the error "Invalid modifier specified: $unset". Could it be that tokens.readAndWrite does not exist? In the error message it also refers to "tokens.readAndWrite_1". I tried the same with the underscore-one at the end, which still didn't work. What am I missing here?

**Comment by ShaneKilkelly** [comment]
Ah, silly me. The correct way to use `$unset` is:

```
db.projects.update({_id: ObjectId("...")}, {$unset: {'tokens.readAndWrite': 1}})
```

**Comment by marcuslindner** [comment]
This command works (in some way). It doesn't give me any feedback at the mongodb shell but it also doesn't seem to have any effect. When I check afterwards again with the command "db.projects.find({'tokens.readAndWrite': null},{'tokens.readAndWrite':1})", I get exactly the same output as before. Starting the docker container shows also no effect. The logs show the same error and the editor also doesn't come up when I open a project.

**Comment by ShaneKilkelly** [comment]
Ah, I think you'll need to add `, {multi: true}` as the last argument to the command:

```
db.projects.update({_id: ObjectId("...")}, {$unset: {'tokens.readAndWrite': 1}}, {multi: true})
```

**Comment by marcuslindner** [comment]
I'm so sorry for bothering you again but I'm not familiar with mongodb. This command doesn't report any error but also has no effect. At least, no observable effect for me. It all remains the same after executing this command and the null entry error is still reported.

**Comment by ShaneKilkelly** [comment]
Hi @marcuslindner , 

That's a shame :( Here's a few things to check:

- Be sure that you're running those commands against the `sharelatex` database each time (Type `use sharelatex` in the prompt first)
- Run a `db.projects.count()` to check that the thing you're operating on actually has project records in it. 

Beyond that, this may not be a problem we can help with very much. I'd recommend checking out the Getting Started guide to MongoDB https://docs.mongodb.com/manual/tutorial/getting-started/ , to get some familiarity with MongoDB.

**Comment by marcuslindner** [comment]
Hi @ShaneKilkelly,

I very much understand that. I checked all these points and I'm on the right database and there is also projects. That's not a problem.

Our DB version is 2.4.10 and too old, as you already mentioned. We have to update, obviously. I will try to update the database and use mongodump (as explained on your wiki pages) and see if that works.

**Comment by marcuslindner** [comment]
I tried now to backup the content of the old database with mongodump. That worked (as far as I can see). Then, I started on a new computer the docker images with your provided docker-compose file.

When I try to restore the mongodb content, I get a lot of errors like this:
![image](https://user-images.githubusercontent.com/8193090/69906819-c4d6ff00-13c9-11ea-88bf-101abd30c021.png)

At the end, I get the final information that "0 document(s) restored successfully. 432 document(s) failed to restore."

Maybe, you have a hint for me?
